community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java

```diff
@@ -44,129 +61,243 @@ public CommunityLockClient( LockManagerImpl manager )
+        }
     }
 
     @Override
     public void releaseShared( Locks.ResourceType resourceType, long resourceId )
     {
-        PrimitiveLongObjectMap<LockResource> localLocks = localShared( resourceType );
-        LockResource resource = localLocks.get( resourceId );
-        if( resource.releaseReference() != 0)
+        if ( !stateHolder.incrementActiveClients() )
         {
             return;
         }
+        try
+        {
+            PrimitiveLongObjectMap<LockResource> localLocks = localShared( resourceType );
+            LockResource resource = localLocks.get( resourceId );
+            if ( resource.releaseReference() != 0 )
+            {
+                return;
+            }
 
-        localLocks.remove( resourceId );
+            localLocks.remove( resourceId );
 
-        manager.releaseReadLock( new LockResource( resourceType, resourceId ), lockTransaction );
+            manager.releaseReadLock( new LockResource( resourceType, resourceId ), lockTransaction );
+        }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
     }
 
     @Override
     public void releaseExclusive( Locks.ResourceType resourceType, long resourceId )
     {
-        PrimitiveLongObjectMap<LockResource> localLocks = localExclusive( resourceType );
-        LockResource resource = localLocks.get( resourceId );
-        if( resource.releaseReference() != 0)
+        if ( !stateHolder.incrementActiveClients() )
         {
             return;
         }
-        localLocks.remove( resourceId );
+        try
+        {
+            PrimitiveLongObjectMap<LockResource> localLocks = localExclusive( resourceType );
+            LockResource resource = localLocks.get( resourceId );
+            if ( resource.releaseReference() != 0 )
+            {
+                return;
+            }
+            localLocks.remove( resourceId );
 
-        manager.releaseWriteLock( new LockResource( resourceType, resourceId ), lockTransaction );
+            manager.releaseWriteLock( new LockResource( resourceType, resourceId ), lockTransaction );
+        }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
     }
 
     @Override
     public void releaseAll()
     {
+        if ( !stateHolder.incrementActiveClients() )
+        {
+            return;
+        }
+        try
+        {
+            releaseLocks();
+        }
+        finally
+        {
+            stateHolder.decrementActiveClients();
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        // closing client to prevent any new client to come
+        stateHolder.closeClient();
+        // wake up and terminate waiters
+        terminateAllWaiters();
+        // wait for all active clients to go and terminate latecomers
+        while ( stateHolder.hasActiveClients() )
+        {
+            terminateAllWaiters();
+            LockSupport.parkNanos( TimeUnit.MILLISECONDS.toNanos( 20 ) );
```

Comment:
I don't know if you have measured this, but I would suspect that waiting just 10 milliseconds would give you better responsiveness with an imperceptible increase in CPU usage.